README.md

## Datos y herramientas `elector`ales de `Ar`rgentina usando `R` <a><img src="man/figures/logo.png" width="200" align="right" /></a>

`{electorAr}` brinda herramientas que facilitan el acceso y el trabajo
con datos electorales de Argentina desde `R`.

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/electorAr)](https://CRAN.R-project.org/package=electorAr)
[![R-CMD-check](https://github.com/PoliticaArgentina/electorAr/workflows/R-CMD-check/badge.svg)](https://github.com/PoliticaArgentina/electorAr/actions)

<!-- badges: end -->

------------------------------------------------------------------------

### INSTALACIÓN

### Versión en desarrollo (*Development version*)

``` r
# install.packages('devtools') si no tiene instalado devtools

devtools::install_github("politicaargentina/electorAr")
```

## Ejemplo de uso (*Usage*)

``` r
library(electorAr)

# Elecciones disponibles según dos fuente de datos
# 'results' para datos agregados y 'data' para escrutinios provisorios 

show_available_elections(source = "results")
```

    ## # A tibble: 728 x 5
    ##    district category round  year  NOMBRE   
    ##    <chr>    <chr>    <chr>  <chr> <chr>    
    ##  1 arg      presi    balota 2015  ARGENTINA
    ##  2 arg      presi    gral   1946  ARGENTINA
    ##  3 arg      presi    gral   1951  ARGENTINA
    ##  4 arg      presi    gral   1958  ARGENTINA
    ##  5 arg      presi    gral   1963  ARGENTINA
    ##  6 arg      presi    gral   1983  ARGENTINA
    ##  7 arg      presi    gral   1989  ARGENTINA
    ##  8 arg      presi    gral   1995  ARGENTINA
    ##  9 arg      presi    gral   1999  ARGENTINA
    ## 10 arg      presi    gral   2003  ARGENTINA
    ## # ... with 718 more rows

``` r
# Filtro eleciones de interés
(gobernadores_tucuman <- show_available_elections(source = "results") %>% 
  dplyr::filter(NOMBRE == "TUCUMAN", 
                category == "gober"))
```

    ## # A tibble: 10 x 5
    ##    district category round year  NOMBRE 
    ##    <chr>    <chr>    <chr> <chr> <chr>  
    ##  1 tucuman  gober    gral  1983  TUCUMAN
    ##  2 tucuman  gober    gral  1987  TUCUMAN
    ##  3 tucuman  gober    gral  1991  TUCUMAN
    ##  4 tucuman  gober    gral  1995  TUCUMAN
    ##  5 tucuman  gober    gral  1999  TUCUMAN
    ##  6 tucuman  gober    gral  2003  TUCUMAN
    ##  7 tucuman  gober    gral  2007  TUCUMAN
    ##  8 tucuman  gober    gral  2011  TUCUMAN
    ##  9 tucuman  gober    gral  2015  TUCUMAN
    ## 10 tucuman  gober    gral  2019  TUCUMAN

``` r
# Descargo multiples elecciones seleccionadas

(elecciones <- get_multiple_elections(gobernadores_tucuman,
                                      source = 'results' ))
```

    ## # A tibble: 10 x 2
    ## # Groups:   id [10]
    ##    id                      election           
    ##    <glue>                  <list>             
    ##  1 tucuman_gober_gral_1983 <spc_tbl_ [19 x 9]>
    ##  2 tucuman_gober_gral_1987 <spc_tbl_ [16 x 9]>
    ##  3 tucuman_gober_gral_1991 <spc_tbl_ [12 x 8]>
    ##  4 tucuman_gober_gral_1995 <spc_tbl_ [11 x 8]>
    ##  5 tucuman_gober_gral_1999 <spc_tbl_ [9 x 8]> 
    ##  6 tucuman_gober_gral_2003 <spc_tbl_ [10 x 8]>
    ##  7 tucuman_gober_gral_2007 <spc_tbl_ [15 x 8]>
    ##  8 tucuman_gober_gral_2011 <spc_tbl_ [17 x 8]>
    ##  9 tucuman_gober_gral_2015 <spc_tbl_ [9 x 8]> 
    ## 10 tucuman_gober_gral_2019 <spc_tbl_ [11 x 8]>

``` r
# Seleciono la elección a Gobernador de TUCUMAN en 1991
(eleccion91 <- elecciones$election[[3]])
```

    ## # A tibble: 12 x 8
    ##    category round  year codprov name_prov listas                   votos elect~1
    ##    <chr>    <chr> <int> <chr>   <chr>     <chr>                    <dbl>   <dbl>
    ##  1 gober    gral   1991 23      TUCUMAN   Frente De La Esperanza  284479  771330
    ##  2 gober    gral   1991 23      TUCUMAN   Fuerza Republicana      247802  771330
    ##  3 gober    gral   1991 23      TUCUMAN   Union Civica Radical     23424  771330
    ##  4 gober    gral   1991 23      TUCUMAN   Laborista De Tucuman      3382  771330
    ##  5 gober    gral   1991 23      TUCUMAN   Frente Por El Progreso~    886  771330
    ##  6 gober    gral   1991 23      TUCUMAN   Obrero                     755  771330
    ##  7 gober    gral   1991 23      TUCUMAN   Movimiento Al Socialis~    724  771330
    ##  8 gober    gral   1991 23      TUCUMAN   Comunista                  511  771330
    ##  9 gober    gral   1991 23      TUCUMAN   Federal                    460  771330
    ## 10 gober    gral   1991 23      TUCUMAN   Defensa Provincial Ban~    325  771330
    ## 11 gober    gral   1991 23      TUCUMAN   Votos En Blanco           9145  771330
    ## 12 gober    gral   1991 23      TUCUMAN   Votos Nulos               3162  771330
    ## # ... with abbreviated variable name 1: electores

``` r
# Calculo indicadores

# Competitividad
compute_competitiveness(eleccion91)
```

    ## # A tibble: 1 x 5
    ##   codprov competitividad  year category round
    ##   <chr>            <dbl> <int> <chr>    <chr>
    ## 1 23               0.936  1991 gober    gral

``` r
# Concentración
compute_concentration(eleccion91)
```

    ## # A tibble: 1 x 5
    ##   codprov concentration  year category round
    ##   <chr>           <dbl> <int> <chr>    <chr>
    ## 1 23               0.93  1991 gober    gral

``` r
# Número Efectivo de Partidos (NEP)
compute_nep(eleccion91)
```

    ## # A tibble: 2 x 6
    ##   value index             year category round codprov
    ##   <dbl> <chr>            <int> <chr>    <chr> <chr>  
    ## 1  2.16 Golosov           1991 gober    gral  23     
    ## 2  2.31 Laakso-Taagepera  1991 gober    gral  23

## `{electorAr}` es parte del universo de paquetes **polAr**

![](https://github.com/PoliticaArgentina/data_warehouse/raw/master/hex/collage.png)<!-- -->


NEWS.md
# electorAr 0.0.1.3

Fixes #5 BUG: `show_available_elections()` now with a new worflow (not scrapping and interacting with Github API)

# electorAr 0.0.1.2

* Fixes #4 fix BUG: `NA` generation y San Luis elections when calling  `get_election_results()` related to #2

# electorAr 0.0.1.1

* Fixes #2 hacking San Luis district values in `show_available_elections()` & `get_election_results()`

# electorAr 0.0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/get_election_results.R

#' Descarga resultados electorales
#' (\emph{Download electoral results})
#'
#' @description
#' Función que descarga resultados electorales agregados, nacionales y provinciales desde 2007
#'
#'  (\emph{Function that downloads national electoral data since 2007})
#'
#' @param district un character con código para Argentina y las 24 provincias
#'
#'  (\emph{a named character with code for Argentina and the 24 provinces}).
#'
#' @param category un character para la categoría electoral: diputado \code{dip}, senador  \code{sen} o  presidente \code{presi}
#'
#'  (\emph{a character with a name for the electoral category: deputy \code{dip}, senator \code{sen} or president \code{presi}}).
#'
#' @param year un integer para el año de eleccion
#'
#' (\emph{an integer for the year of choice}).
#'
#' @param round un character para tipo de elección: primaria \code{paso} o general  \code{gral}
#'
#'  (\emph{a character with a name for the election round: primary \code{paso} or general \code{gral}}).
#'
#'@return devuelve un tibble con \code{class "spec_tbl_df", "tbl_df","tbl", "data.frame"} con los resultados de una elección determinada
#'
#'  (\emph{returns a tibble with electoral results of interest with \code{class "spec_tbl_df", "tbl_df","tbl", "data.frame"}}).
#'
#' @examples
#'  get_election_data(district = "caba",
#'                    category = "dip",
#'                    round = "paso",
#'                    year = 2011,
#'                    level = "provincia",
#'                    long = TRUE, raw = FALSE)
#' @export

get_election_results <- function(district = NULL ,
                                 category = NULL,
                                 round = NULL ,
                                 year = NULL){


  ## Check for internet conection
  attempt::stop_if_not(.x = curl::has_internet(),
                       msg = "Internet access was not detected. Please check your connection //
No se detecto acceso a internet. Por favor chequear la conexion.")

  ## Check params

  # year check

  assertthat::assert_that(!is.null(year),
                          msg = "You must provide valid character parameters for 'district', 'category' and 'round', and an integer parameter for 'year'. Check them with 'show_available_elections()' //
Debe proporcionar parametros character validos para 'district', 'category' y 'round', y un parametro entero para 'year'. Compruebelos con 'show_available_elections()'")

  assertthat::assert_that(is.numeric(year),
                          msg = "'year' must be an integer. Please select a correct option.  Check them with 'show_available_elections(source = 'results')' //
'year' debe ser un integer. Por favor seleccione una opcipn correcta. Verifiquelos con 'show_available_elections(source = 'results')'")

  # category check

  assertthat::assert_that(!is.null(category),
                          msg = "You must provide valid character parameters for 'district', 'category' and 'round', and an integer parameter for 'year'. Check them with show 'show_available_elections()' //
 Debe proporcionar parametros character validos para 'district', 'category' y 'round', y un parametro entero para 'year'. Compruebelos con 'show show_available_elections()'")

  assertthat::assert_that(is.character(category),
                          msg = "'category' must be an character string Please select a correct option.  Check them with 'show_available_elections(source = 'results')' //
'category' debe ser un character string Por favor seleccione una opcipn correcta. Verifiquelas con 'show_available_elections(source = 'results')'")

  assertthat::assert_that(category %in% c("gober", "presi"),
                          msg = "Please select a correct 'category'. Check them with 'show_available_elections(source = 'results')' //
Por favor seleccione una 'catgeory' correcta. Compruebelas con 'show_available_elections(source = 'results')'")


  # district check

  assertthat::assert_that(!is.null(district),
                          msg = "You must provide valid character parameters for 'district', 'category' and 'round', and an integer parameter for 'year'. Check them with 'show_available_elections()' //
Debe proporcionar parametros character validos para 'district', 'category' y 'round', y un parametro integer para 'year'. Compruebelos con  'show_available_elections()'")



  assertthat::assert_that(is.character(district),
                          msg = "'district' must be a character string. Check options with 'show_available_elections(source = 'results')' //
Por favor seleccione un 'district' correcto. Compruebelos con 'show_available_elections(source = 'results')')")

  assertthat::assert_that(district %in% c("arg",
                                          "caba",
                                          "catamarca",
                                          "chaco",
                                          "chubut",
                                          "cordoba",
                                          "corrientes",
                                          "erios",
                                          "formosa",
                                          "jujuy",
                                          "mendoza",
                                          "misiones",
                                          "neuquen",
                                          "pampa",
                                          "pba",
                                          "rioja",
                                          "rnegro",
                                          "salta",
                                          "santiago",
                                          "scruz",
                                          "sfe",
                                          "sjuan",
                                          "sluis",
                                          "tdf",
                                          "tucuman"),
                          msg = "Please select a correct 'district.' Check them with 'show_available_elections()' //
Por favor seleccione un 'district' correcto. Compruebelos con 'show_available_elections(source = 'results')")

  # round check

  assertthat::assert_that(!is.null(round),
                          msg = "You must provide valid character parameters for 'district', 'category' and 'round', and an integer parameter for 'year'. Check them with 'show_available_elections()' //
Debe proporcionar parametros character validos para 'district', 'category' y 'round', y un parametro integer para 'year'. Compruebelos con  'show_available_elections()'")


  assertthat::assert_that(is.character(round),
                          msg = "'round' must be a character string. Options = c('paso', 'gral', 'balota') //
'round' debe ser un character string. Opciones = c('paso', 'gral', 'balota')")

  assertthat::assert_that(round %in% c("paso", "gral", "balota"),
                          msg = "Please select a correct 'round' Check them with 'show_available_elections(source = 'results')' //
Por favor seleccione una 'round' correcta. Compruebelas con 'show_available_elections(source = 'results')'")


  # Check available elections


  check_elections <-   electorAr::show_available_elections(source = 'results') %>%
    dplyr::mutate(elections =paste0(district, category, round, year))


  assertthat::assert_that(glue::glue({district}, {category}, {round}, {year}) %in% check_elections$elections,
                          msg = "Please choose a valid election. Check them with 'show_available_elections(source = 'results')' //
Por favor seleccione una eleccipn valida. Consultelas con 'show_available_elections(source = 'results')'")


  ### Provincial Electoral codes # INDRA


  codProv <- tibble::tibble(name_prov =
                              c("CABA", "CATAMARCA", "CHACO", "CHUBUT", "CORDOBA", "CORRIENTES","ENTRE RIOS","FORMOSA",
                                "JUJUY", "LA PAMPA", "LA RIOJA", "MENDOZA", "MISIONES", "NEUQUEN","BUENOS AIRES","RIO NEGRO",
                                "SALTA", "SANTA CRUZ", "SANTA FE", "SANTIAGO DEL ESTERO", "SAN JUAN", "SAN LUIS", "TIERRA DEL FUEGO", "TUCUMAN"),
                            codprov =
                              c('01','03','06','07','04','05','08','09',
                                '10','11','12','13','14','15','02','16',
                                '17','20','21','22','18','19','24','23'))


    ### SAN LUIS DISTRICT HARDCODING

  district <- if(district == "sluis"){

       "sanluis"} else{

         district
     }

            ## ELECTION URL
            url <- paste0("https://raw.githubusercontent.com/PoliticaArgentina/data_warehouse/master/electorAr/data/escrutinios_definitivos/",
                          district, "_",
                          category, "_",
                          round,
                          year, ".csv")

            # GET DATA

            # Set default value for try()

            default <- NULL

            df <- base::suppressWarnings(base::try(default <- vroom::vroom(file = url,
                                                                           col_types = vroom::cols(),
                                                                           progress = FALSE),
                                                   silent = TRUE))


            # FAIL SAFELY WITH ERROR MESSAGE
            if(is.null(default)){

              df <- base::message("Fail to download data. Source is not available // La fuente de datos no esta disponible")

            } else {


              df <- df %>%
                dplyr::mutate(category = category,
                              round = round,
                              year = year,
                              district = district)

              #add district names from districts id

              df <-  df %>%
                dplyr::mutate(name_prov = dplyr::case_when(
                  district =="arg" ~    "argentina",
                  district =="caba" ~     "caba",
                  district =="catamarca" ~ "catamarca",
                  district =="chaco" ~    "chaco",
                  district =="chubut" ~   "chubut",
                  district =="cordoba" ~  "cordoba",
                  district =="corrientes" ~ "corrientes",
                  district =="erios" ~    "entre rios",
                  district =="formosa" ~  "formosa",
                  district =="jujuy" ~    "jujuy",
                  district =="mendoza" ~  "mendoza",
                  district =="misiones" ~ "misiones",
                  district =="neuquen" ~  "neuquen",
                  district =="pampa" ~  "la pampa",
                  district =="pba" ~    "buenos aires",
                  district =="rioja" ~  "la rioja",
                  district =="rnegro" ~   "rio negro",
                  district =="salta" ~  "salta",
                  district =="santiago" ~ "santiago del estero",
                  district =="scruz" ~  "santa cruz",
                  district =="sfe" ~    "santa fe",
                  district =="sjuan" ~  "san juan",
                  district =="sanluis" ~  "san luis",
                  district =="tdf" ~  "tierra del fuego",
                  district =="tucuman" ~ "tucuman")) %>%
                dplyr::mutate(name_prov = stringr::str_to_upper(name_prov)) %>%
                dplyr::select(-district)


              # JOIN DISTRCIT CODE

               df <-  df %>%
                dplyr::left_join(codProv, by = "name_prov") %>%
                dplyr::rename(listas = lista) %>%
                dplyr::select(category, round, year, codprov, name_prov, dplyr::everything())

            }

            return(df)

  } # CLOSE FUNCTION

